### ISSUE TITLE
[CI] Netty Log Appender IOException

### ISSUE THREAD

**Issue by albertzaharovits** [elastic/elasticsearch#27306]
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+oracle-periodic/1401/console
Does not reproduce locally:
```
gradle :modules:transport-netty4:integTestRunner \
  -Dtests.seed=796922BD0ED8F959 \
  -Dtests.class=org.elasticsearch.transport.netty4.ESLoggingHandlerIT \
  -Dtests.method="testLoggingHandler" \
  -Dtests.security.manager=true \
  -Dtests.locale=sl-SI \
  -Dtests.timezone=Asia/Hong_Kong
```
Multiple:
```
Connection reset by peer
  1> java.io.IOException: Connection reset by peer
  1> 	at sun.nio.ch.FileDispatcherImpl.read0(Native Method) ~[?:?]
  1> 	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39) ~[?:?]
  1> 	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223) ~[?:?]
  1> 	at sun.nio.ch.IOUtil.read(IOUtil.java:197) ~[?:?]
  1> 	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380) ~[?:?]
  1> 	at io.netty.buffer.PooledHeapByteBuf.setBytes(PooledHeapByteBuf.java:261) ~[netty-buffer-4.1.13.Final.jar:4.1.13.Final]
  1> 	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1100) ~[netty-buffer-4.1.13.Final.jar:4.1.13.Final]
  1> 	at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:372) ~[netty-transport-4.1.13.Final.jar:4.1.13.Final]
  1> 	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:123) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
  1> 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
  1> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:544) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
  1> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:498) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
  1> 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
  1> 	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858) [netty-common-4.1.13.Final.jar:4.1.13.Final]
  1> 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
```
Causing: 
```
at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.elasticsearch.test.MockLogAppender$PatternSeenEventExcpectation.assertMatched(MockLogAppender.java:153)
	at org.elasticsearch.test.MockLogAppender.assertAllExpectationsMatched(MockLogAppender.java:62)
	at org.elasticsearch.transport.netty4.ESLoggingHandlerIT.testLoggingHandler(ESLoggingHandlerIT.java:80)
```

OS | CentOS 7.4.1708, Linux 3.10.0-693.5.2.el7.x86_64
-- | --
Specs | 4 CPUs, 15.51GB RAM
java.version | 1.8.0_144
java.vm.name | Java HotSpot(TM) 64-Bit Server VM
java.vm.version | 25.144-b01
java.runtime.version | 1.8.0_144-b01

**Comment by cbuescher** [comment]
Another one today: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+multijob-unix-compatibility/os=oraclelinux/516/console

**Comment by cbuescher** [comment]
Another one today, looks like it has some interesting details about the offending character in the trace message:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.6+periodic/1376/console

```
11:29:28   1> [2018-01-29T03:29:24,266][TRACE][o.e.t.n.ESLoggingHandler ] an exception occurred formatting a trace message
11:29:28   1> java.io.IOException: Invalid string; unexpected character: 191 hex: bf
11:29:28   1> 	at org.elasticsearch.common.io.stream.StreamInput.readString(StreamInput.java:372) ~[elasticsearch-5.6.7-SNAPSHOT.jar:5.6.7-SNAPSHOT]
11:29:28   1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ThreadContextStruct.<init>(ThreadContext.java:362) ~[elasticsearch-5.6.7-SNAPSHOT.jar:5.6.7-SNAPSHOT]
11:29:28   1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ThreadContextStruct.<init>(ThreadContext.java:352) ~[elasticsearch-5.6.7-SNAPSHOT.jar:5.6.7-SNAPSHOT]
11:29:28   1> 	at org.elasticsearch.common.util.concurrent.ThreadContext.readHeaders(ThreadContext.java:186) ~[elasticsearch-5.6.7-SNAPSHOT.jar:5.6.7-SNAPSHOT]
11:29:28   1> 	at org.elasticsearch.transport.netty4.ESLoggingHandler.format(ESLoggingHandler.java:105) ~[main/:?]
11:29:28   1> 	at org.elasticsearch.transport.netty4.ESLoggingHandler.format(ESLoggingHandler.java:50) [main/:?]
11:29:28   1> 	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:239) [netty-handler-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:134) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:644) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:544) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:498) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:458) [netty-transport-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858) [netty-common-4.1.13.Final.jar:4.1.13.Final]
11:29:28   1> 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
11:29:28   1> [2018-01-29T03:29:24,277][TRACE][o.e.t.n.ESLoggingHandler ] [id: 0x4b87033f, L:/127.0.0.1:42022 - R:/127.0.0.1:39680] READ: 1097B
11:29:28   1>          +-------------------------------------------------+
11:29:28   1>          |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
11:29:28   1> +--------+-------------------------------------------------+----------------+
11:29:28   1> |00000000| 45 53 ff ff ff ff 45 53 00 00 04 3d 00 00 00 00 |ES....ES...=....|
11:29:28   1> |00000010| 00 00 00 10 00 00 4d 38 bf 00 01 07 57 61 72 6e |......M8....Warn|
11:29:28   1> |00000020| 69 6e 67 03 83 02 32 39 39 20 45 6c 61 73 74 69 |ing...299 Elasti|
11:29:28   1> |00000030| 63 73 65 61 72 63 68 2d 35 2e 36 2e 37 2d 53 4e |csearch-5.6.7-SN|
11:29:28   1> |00000040| 41 50 53 48 4f 54 2d 37 66 61 64 39 32 39 20 22 |APSHOT-7fad929 "|
11:29:28   1> |00000050| 5b 73 63 72 69 70 74 2e 6d 61 78 5f 63 6f 6d 70 |[script.max_comp|
11:29:28   1> |00000060| 69 6c 61 74 69 6f 6e 73 5f 70 65 72 5f 6d 69 6e |ilations_per_min|
11:29:28   1> |00000070| 75 74 65 5d 20 73 65 74 74 69 6e 67 20 77 61 73 |ute] setting was|
11:29:28   1> |00000080| 20 64 65 70 72 65 63 61 74 65 64 20 69 6e 20 45 | deprecated in E|
11:29:28   1> |00000090| 6c 61 73 74 69 63 73 65 61 72 63 68 20 61 6e 64 |lasticsearch and|
11:29:28   1> |000000a0| 20 77 69 6c 6c 20 62 65 20 72 65 6d 6f 76 65 64 | will be removed|
11:29:28   1> |000000b0| 20 69 6e 20 61 20 66 75 74 75 72 65 20 72 65 6c | in a future rel|
11:29:28   1> |000000c0| 65 61 73 65 21 20 53 65 65 20 74 68 65 20 62 72 |ease! See the br|
11:29:28   1> |000000d0| 65 61 6b 69 6e 67 20 63 68 61 6e 67 65 73 20 64 |eaking changes d|
11:29:28   1> |000000e0| 6f 63 75 6d 65 6e 74 61 74 69 6f 6e 20 66 6f 72 |ocumentation for|
11:29:28   1> |000000f0| 20 74 68 65 20 6e 65 78 74 20 6d 61 6a 6f 72 20 | the next major |
11:29:28   1> |00000100| 76 65 72 73 69 6f 6e 2e 22 20 22 4d 6f 6e 2c 20 |version." "Mon, |
11:29:28   1> |00000110| 32 39 20 4a 61 6e 20 32 30 31 38 20 31 31 3a 32 |29 Jan 2018 11:2|
11:29:28   1> |00000120| 39 3a 32 32 20 47 4d 54 22 ee 01 32 39 39 20 45 |9:22 GMT"..299 E|
11:29:28   1> |00000130| 6c 61 73 74 69 63 73 65 61 72 63 68 2d 35 2e 36 |lasticsearch-5.6|
11:29:28   1> |00000140| 2e 37 2d 53 4e 41 50 53 48 4f 54 2d 37 66 61 64 |.7-SNAPSHOT-7fad|
11:29:28   1> |00000150| 39 32 39 20 22 5b 73 63 72 69 70 74 2e 69 6e 6c |929 "[script.inl|
11:29:28   1> |00000160| 69 6e 65 5d 20 73 65 74 74 69 6e 67 20 77 61 73 |ine] setting was|
11:29:28   1> |00000170| 20 64 65 70 72 65 63 61 74 65 64 20 69 6e 20 45 | deprecated in E|
11:29:28   1> |00000180| 6c 61 73 74 69 63 73 65 61 72 63 68 20 61 6e 64 |lasticsearch and|
11:29:28   1> |00000190| 20 77 69 6c 6c 20 62 65 20 72 65 6d 6f 76 65 64 | will be removed|
11:29:28   1> |000001a0| 20 69 6e 20 61 20 66 75 74 75 72 65 20 72 65 6c | in a future rel|
11:29:28   1> |000001b0| 65 61 73 65 21 20 53 65 65 20 74 68 65 20 62 72 |ease! See the br|
11:29:28   1> |000001c0| 65 61 6b 69 6e 67 20 63 68 61 6e 67 65 73 20 64 |eaking changes d|
11:29:28   1> |000001d0| 6f 63 75 6d 65 6e 74 61 74 69 6f 6e 20 66 6f 72 |ocumentation for|
11:29:28   1> |000001e0| 20 74 68 65 20 6e 65 78 74 20 6d 61 6a 6f 72 20 | the next major |
11:29:28   1> |000001f0| 76 65 72 73 69 6f 6e 2e 22 20 22 4d 6f 6e 2c 20 |version." "Mon, |
11:29:28   1> |00000200| 32 39 20 4a 61 6e 20 32 30 31 38 20 31 31 3a 32 |29 Jan 2018 11:2|
11:29:28   1> |00000210| 39 3a 32 32 20 47 4d 54 22 ee 01 32 39 39 20 45 |9:22 GMT"..299 E|
11:29:28   1> |00000220| 6c 61 73 74 69 63 73 65 61 72 63 68 2d 35 2e 36 |lasticsearch-5.6|
11:29:28   1> |00000230| 2e 37 2d 53 4e 41 50 53 48 4f 54 2d 37 66 61 64 |.7-SNAPSHOT-7fad|
11:29:28   1> |00000240| 39 32 39 20 22 5b 73 63 72 69 70 74 2e 73 74 6f |929 "[script.sto|
11:29:28   1> |00000250| 72 65 64 5d 20 73 65 74 74 69 6e 67 20 77 61 73 |red] setting was|
11:29:28   1> |00000260| 20 64 65 70 72 65 63 61 74 65 64 20 69 6e 20 45 | deprecated in E|
11:29:28   1> |00000270| 6c 61 73 74 69 63 73 65 61 72 63 68 20 61 6e 64 |lasticsearch and|
11:29:28   1> |00000280| 20 77 69 6c 6c 20 62 65 20 72 65 6d 6f 76 65 64 | will be removed|
11:29:28   1> |00000290| 20 69 6e 20 61 20 66 75 74 75 72 65 20 72 65 6c | in a future rel|
11:29:28   1> |000002a0| 65 61 73 65 21 20 53 65 65 20 74 68 65 20 62 72 |ease! See the br|
11:29:28   1> |000002b0| 65 61 6b 69 6e 67 20 63 68 61 6e 67 65 73 20 64 |eaking changes d|
11:29:28   1> |000002c0| 6f 63 75 6d 65 6e 74 61 74 69 6f 6e 20 66 6f 72 |ocumentation for|
11:29:28   1> |000002d0| 20 74 68 65 20 6e 65 78 74 20 6d 61 6a 6f 72 20 | the next major |
11:29:28   1> |000002e0| 76 65 72 73 69 6f 6e 2e 22 20 22 4d 6f 6e 2c 20 |version." "Mon, |
11:29:28   1> |000002f0| 32 39 20 4a 61 6e 20 32 30 31 38 20 31 31 3a 32 |29 Jan 2018 11:2|
11:29:28   1> |00000300| 39 3a 32 32 20 47 4d 54 22 24 63 6c 75 73 74 65 |9:22 GMT"$cluste|
11:29:28   1> |00000310| 72 3a 6d 6f 6e 69 74 6f 72 2f 6e 6f 64 65 73 2f |r:monitor/nodes/|
11:29:28   1> |00000320| 68 6f 74 5f 74 68 72 65 61 64 73 5b 6e 5d 16 58 |hot_threads[n].X|
11:29:28   1> |00000330| 4d 4d 74 55 34 32 67 53 73 4b 30 4c 33 56 37 65 |MMtU42gSsK0L3V7e|
11:29:28   1> |00000340| 36 44 72 49 77 00 00 00 00 00 00 00 0f 16 59 53 |6DrIw.........YS|
11:29:28   1> |00000350| 73 7a 6f 35 72 77 52 62 71 53 4e 50 65 52 76 32 |szo5rwRbqSNPeRv2|
11:29:28   1> |00000360| 51 4a 5f 77 00 00 01 02 07 6e 6f 64 65 5f 73 31 |QJ_w.....node_s1|
11:29:28   1> |00000370| 16 58 4d 4d 74 55 34 32 67 53 73 4b 30 4c 33 56 |.XMMtU42gSsK0L3V|
11:29:28   1> |00000380| 37 65 36 44 72 49 77 16 58 5a 6e 73 6c 46 31 35 |7e6DrIw.XZnslF15|
11:29:28   1> |00000390| 52 39 43 43 6c 50 4c 5f 39 4c 33 73 62 77 09 31 |R9CClPL_9L3sbw.1|
11:29:28   1> |000003a0| 32 37 2e 30 2e 30 2e 31 09 31 32 37 2e 30 2e 30 |27.0.0.1.127.0.0|
11:29:28   1> |000003b0| 2e 31 00 01 04 7f 00 00 01 09 31 32 37 2e 30 2e |.1........127.0.|
11:29:28   1> |000003c0| 30 2e 31 00 00 83 77 00 03 00 01 02 bf f1 b4 02 |0.1...w.........|
11:29:28   1> |000003d0| 07 6e 6f 64 65 5f 73 30 16 59 53 73 7a 6f 35 72 |.node_s0.YSszo5r|
11:29:28   1> |000003e0| 77 52 62 71 53 4e 50 65 52 76 32 51 4a 5f 77 16 |wRbqSNPeRv2QJ_w.|
11:29:28   1> |000003f0| 4a 50 67 63 72 33 55 38 52 34 36 55 5f 39 64 6d |JPgcr3U8R46U_9dm|
11:29:28   1> |00000400| 4b 74 4e 6b 57 51 09 31 32 37 2e 30 2e 30 2e 31 |KtNkWQ.127.0.0.1|
11:29:28   1> |00000410| 09 31 32 37 2e 30 2e 30 2e 31 00 01 04 7f 00 00 |.127.0.0.1......|
11:29:28   1> |00000420| 01 09 31 32 37 2e 30 2e 30 2e 31 00 00 a4 26 00 |..127.0.0.1...&.|
11:29:28   1> |00000430| 03 00 01 02 bf f1 b4 02 00 00 00 00 03 01 03 63 |...............c|
11:29:28   1> |00000440| 70 75 e8 07 02 00 00 00 0a                      |pu.......       |
11:29:28   1> +--------+-------------------------------------------------+----------------+
```

**Comment by tbrooks8** [comment]
I think this is a bug in the `ESLoggingHandler`. In order to handle `pings` we check if the message is 6 bytes long.

```
if (readableBytes == TcpHeader.MARKER_BYTES_SIZE + TcpHeader.MESSAGE_LENGTH_SIZE) {
                    final int length = arg.getInt(offset + MESSAGE_LENGTH_OFFSET);
                    if (length == TcpTransport.PING_DATA_SIZE) {
                        sb.append(" [ping]").append(' ').append(eventName).append(": ").append(readableBytes).append('B');
                        return sb.toString();
                    }
                }
```

But in the case above, a ping and a follow message message has been read in one go. And therefore we a mishandling the ping. I will look into a PR to fix this.

**Comment by andyb-elastic** [comment]
Another in 5.6. I'll mute this test since Tim's found the cause

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.6+multijob-unix-compatibility/os=centos/692/console

[build-692-ESLoggingHandlerIT.txt](https://github.com/elastic/elasticsearch/files/1729786/build-692-ESLoggingHandlerIT.txt)

**Comment by jasontedor** [comment]
> I will look into a PR to fix this.

Will you open a fix for this?

**Comment by tbrooks8** [comment]
This still need needs to be fixed on 6.x.

**Comment by jasontedor** [comment]
@tbrooks8 Since this represents a bug in production, would you pick this up to fix in 6.x?

**Comment by jasontedor** [comment]
Closed by #32725 and #33789
